src/java.base/share/classes/jdk/internal/module/ModuleBootstrap.java

```diff
@@ -466,7 +464,7 @@ private static ModuleLayer boot2() {
 
         if (CDS.isDumpingStaticArchive()
                 && !haveUpgradeModulePath
-                && addModules.isEmpty()
+                && (addModules.isEmpty() || addModulesFromRuntimeImage(addModules))
```

Comment:
I think it would be better to check the modules in the Configuration as that is what will be archived. Can you try this, calling addJrt(cf, addModules) where this method is below (it's similar to allJrtOrModularJar that we added recently, not tested btw).

```
    /**
     * Returns true if all modules named in the given set are in the Configuration and
     * the run-time image.
     */
    private static boolean allJrt(Configuration cf, Set<String> moduleNames) {
        return !moduleNames.stream()
                .map(mn -> cf.findModule(mn).orElseThrow())
                .map(m -> m.reference().location().orElseThrow())
                .anyMatch(uri -> !uri.getScheme().equalsIgnoreCase("jrt"));
    }
```

Reply:
Hi Alan,
I tried your suggestion but it can't handle the `ALL-SYSTEM` case.
I made some slight adjustments to your patch as follows:

```
    /**
     * Returns true if all modules named in the given set are in the Configuration and
     * the run-time image.
     */
    private static boolean allJrt(Configuration cf, Set<String> moduleNames) {
        if (moduleNames.size() == 1 && moduleNames.contains(ALL_SYSTEM)) {
            return true;
        }
        return !moduleNames.stream()
                .filter(mn -> !mn.equals(ALL_SYSTEM))
                .map(mn -> cf.findModule(mn).orElseThrow())
                .map(m -> m.reference().location().orElseThrow())
                .anyMatch(uri -> !uri.getScheme().equalsIgnoreCase("jrt"));
    }
```
What do you think?

Reply:
Good point as ALL-SYSTEM and ALL-MODULE-PATH are allowed in the value. I assume you don't need to special case the single value case as this is dump time so not performance critical.

Reply:
I chatted with Calvin about this change. Testing that the root modules specified to --add-modules are in the run-time image isn't a sufficient test because these modules, or any modules that they transitively require, may use services that are provided by modules on the application module. An accurate test would require generating a second Configuration without --add-modules and testing that all modules in the set-difference are in the run-time image. That adds a lot of complexity that would be hard to explain and diagnose why dumping works in some cases and not others. Also possible next steps for CDS would allow for modules on the module path so just allowing this now would avoid this complexity.

Reply:
After off-line discussion with Alan, we will allow any module to be included in --add-modules for CDS.